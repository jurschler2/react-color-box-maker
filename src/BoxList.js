import React, { useState } from "react";
import NewBoxForm from "./NewBoxForm";
import Box from "./Box";
import { v4 as uuid } from 'uuid';

function BoxList() {
  const [boxes, setBoxes] = useState([]);

  const renderBoxes = () => {

    return (
      <div className="boxes" data-testid="boxes-container">
        {boxes.map(box => <Box id={box.id} 
                               key={box.id}
                               width={box.width} 
                               height={box.height} 
                               backgroundColor={box.color} 
                               removeBox={removeBox}/>)}
      </div>
    );
  };


  /** Add new box object to the boxlist. */
  const addBox = box => {
    let newBox = { ...box, id: uuid() };
    setBoxes(boxes => [...boxes, newBox]);
  };
  

  /** Remove a box object to the boxlist. */
  const removeBox = (boxID) => {
    
    const boxesCopy = boxes.filter((b)=> b.id !== boxID);
    setBoxes(boxesCopy);
  }

  return (
    <div className="BoxList">
      <NewBoxForm addBox={addBox} />
      {renderBoxes()}
    </div>
  );
};


export default BoxList;