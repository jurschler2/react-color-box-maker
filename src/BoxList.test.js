import React from "react";
import { render, fireEvent } from "@testing-library/react";
import '@testing-library/jest-dom/extend-expect';
import BoxList from "./BoxList";

function addBoxHelper(getByLabelText,queryByText) {

  const widthInput = getByLabelText("Width:");
  const heightInput = getByLabelText("Height:");
  const colorInput = getByLabelText("Color:");
  const submitBox = queryByText("Add a new box!");

  // fill out the form
  fireEvent.change(widthInput, { target: { value: 100 }});
  fireEvent.change(heightInput, { target: { value: 100 }});
  fireEvent.change(colorInput, { target: { value: "red" }});
  fireEvent.click(submitBox);
}

it("renders without crashing", function() {
  render(<BoxList />);
});

it("matches snapshot", function() {
  const {asFragment} = render(<BoxList />);
  expect(asFragment()).toMatchSnapshot();
});

it("can add a new box", function() {
  const { getByLabelText, queryByText, queryByTestId } = render(<BoxList />);

  // no boxes yet
  expect(queryByTestId("boxes-container")).not.toContainHTML("box-element-");

  addBoxHelper(getByLabelText, queryByText);

  // item exists!
  expect(queryByTestId("boxes-container")).toContainHTML("box-element-");
  expect(queryByTestId("boxes-container")).toContainHTML('style="width: 100px; height: 100px; background-color: red;"');
});

it("can delete a box", function() {
  const { getByLabelText, queryByText, queryByTestId } = render(<BoxList />);

  // no boxes yet
  expect(queryByTestId("boxes-container")).not.toContainHTML("box-element-");

  addBoxHelper(getByLabelText, queryByText);

  // item exists!
  expect(queryByTestId("boxes-container")).toContainHTML("box-element-");
  expect(queryByTestId("boxes-container")).toContainHTML('style="width: 100px; height: 100px; background-color: red;"');

  // delete the box
  const deleteBox = queryByText("Delete this box!");
  fireEvent.click(deleteBox);

  // no boxes should be on the page
  expect(queryByTestId("boxes-container")).not.toContainHTML("box-element-");

});