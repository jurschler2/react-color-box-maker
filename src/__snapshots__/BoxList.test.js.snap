// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="BoxList"
  >
    <form>
      <label
        for="width"
      >
        Width:
      </label>
      <input
        id="width"
        name="width"
        value=""
      />
      <label
        for="height"
      >
        Height:
      </label>
      <input
        id="height"
        name="height"
        value=""
      />
      <label
        for="color"
      >
        Color:
      </label>
      <input
        id="color"
        name="color"
        value=""
      />
      <button>
        Add a new box!
      </button>
    </form>
    <div
      class="boxes"
      data-testid="boxes-container"
    />
  </div>
</DocumentFragment>
`;
